Promote LogarithmicScaleDown to Beta

damemi opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This promotes the `LogarithmicScaleDown` feature gate to `beta` and enables it by default.

This was first introduced in 1.21 alpha in https://github.com/kubernetes/kubernetes/pull/99212

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Ref https://github.com/kubernetes/enhancements/issues/2185

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
LogarithmicScaleDown is now Beta and enabled by default
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: https://github.com/kubernetes/enhancements/issues/2185
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/2185
```

damemi reviewed (commented):
/hold
working on updates to the kep

pacoxu commented:
/sig apps

damemi commented:
/hold cancel
@alculquicondor @soltysh could you please take a look at this?
Also pinging @wojtek-t (to review the metric as recommended in the KEP update https://github.com/kubernetes/enhancements/pull/2691)

The metric is a bit tricky, because in order to calculate the ratio of deleted pod's age vs youngest pod's age, we need to find the youngest pod. They aren't necessarily all sorted by age (it is only 1 criteria in the sorting) so I think we need to check them all. This also has the problem that the pods being deleted might not always fall in the `[1,2]` range we expected (because they may be prioritized for their phase rather than age, for example). Hence the logarithmic bucketing is not always guaranteed.

alculquicondor commented on pkg/controller/replicaset/replica_set.go:
```diff
@@ -802,10 +804,28 @@ func getPodsToDelete(filteredPods, relatedPods []*v1.Pod, diff int) []*v1.Pod {
+	}
+
+	// for each pod chosen for deletion, report the ratio of its age to the youngest pod's age
+	for _, pod := range filteredPods[:diff] {
```
In the KEP it says:

```
The metric `deleted_pod_age_ratio` will provide a histogram of the ratio between the 
chosen `deleted pod`'s age over the current `youngest pod`'s age, for pods where the sort 
algorithm falls back to age. (Pod age is the final criteria in the sorting algorithm, so we don't 
want to measure this ratio for deletions which don't use this feature, as those may validly fall 
outside the desired range).
```

However, it seems like it would be hard to tell. So I'm fine with this implementation.

damemi replied:
Yeah, I wanted to make that distinction but looking at the actual implementation we don't have an easy way to single out "what" caused a pod to be sorted in that order.

I do see now you pointed out the metric should be a histogram, so I'll update to that (I just used a gauge)

wojtek-t replied:
So the problem is that currently we can have arbitrary values in the histogram in a valid way.
So that metric won't be useful by any means...

So let's try to make it useful:
- what we really care about are ready pods - so let's focus only on those
- so whenever we delete "ready pod" report its age compared to the youngest "ready" pod.

That solves most of the problems with the metric, modulo these two:
https://github.com/kubernetes/kubernetes/blob/4aae71695a8dd43918702fafd81e0401721d79d9/pkg/controller/controller_utils.go#L836-L851
- the first one we can ignore pods that set that annotation
- the former we can address by computing the youngest within the pods with the same rank
But that would make this metric very hard to intuitively explain to anyone. 

I didn't initially realize that the sorting function has so many different branches. And I'm not sure what to do. Up until we were ignoring non-ready pods, that metric still made some sense. But after that point it's becoming too artificial imho.

Any ideas what to do with this?

alculquicondor replied:
I'm starting to doubt the usefulness of the metric as a whole. If we are removing several pods, it's likely that some of them will be more than twice older than the youngest.

Are we simply assuming that most of the times you would do small scale downs so the number should stay below 2?

If we want to be 100% accurate, we should be grouping by rank. But I think we would be missing an opportunity for the metric to be useful for users beyond checking that the feature works. For example, if the users see that the metric is well above 2, they can start wondering why, and the reason could be that they are not spreading their pods appropriately.

wojtek-t replied:
This was the point I was trying to pass above.

But at the same time, I would really like to be able to answer if the feature works. But given all of the above, we should maybe be looking at the feature as a whole "sorting" function? Or maybe yet a different angle?
I'm looking for ideas what things we may try to report (probably in a wider context) to show that the feature works.

alculquicondor replied:
So the sorting is trying to balance two goals/heuristics:
1. newer pods are generally safer to delete (hasn't warmed up, has done less work, has less open connections, etc).
2. pods should be spread. This is achieved in 2 ways:
  - deleting pods in nodes where there are other pods of the same replicaset
  - random deletion (the feature being graduated).

The proposed metric is trying to make sure we don't bomb 1. Perhaps we should keep it as it is already implemented in this PR.

Now, can we easily report a metric for 2 without increasing runtime significantly? The only information we have at this point is the node name. To make it really useful we should report the zone spread, but then we would have to add a node lister.

damemi replied:
@wojtek-t @alculquicondor in the interest of keeping this on track for code freeze, could I propose a compromise in that we continue with the metric as implemented in this PR, and add a TODO to continue investigating the other metric (possibly for GA promotion?). The conversation seems to have stalled on this one

logicalhan replied:
I was talking it over with @brancz, it may make more sense to have two metrics, from the gist of this conversation (without knowing too much about what you guys are actually doing), it sounds like you may want a `youngest_ready_pod_age` metric and a `deleted_ready_pod_age` metric, or something like this and do the computation after the fact.

alculquicondor replied:
I don't think that would help. The ratio only makes sense at the time of deletion. If you have separate metrics, then you can't match the deleted pod with the "youngest" pod to do the division.

alculquicondor replied:
I think we should be good with the metric as is, except it should be a distribution and only focus on Ready pods, as Wojtek proposed earlier. Any extra filtering sounds like an overkill to me.

alculquicondor replied:
BTW, I'm proposing to get rid of the rank, in the context of #102949

damemi replied:
> The ratio only makes sense at the time of deletion. If you have separate metrics, then you can't match the deleted pod with the "youngest" pod to do the division.

This was my concern with the proposal for 2 metrics as well. Ultimately I think any measurement of this feature will have some level of awkwardness, that may require refinement. I agree that the metric as written is probably the simplest approach for now.

logicalhan replied:
You can use two Gauge Funcs, these will literally evaluate at scrape time, therefore will always expose only current data.

logicalhan replied:
Okay, I talked to @alculquicondor offline, I think I grok what the metric is actually trying to measure. The name is a bit confusing, I think I prefer `sorting_deletion_age_ratio`, since this metric primarily targets the efficacy of the sorting algorithm. I also retract my suggestion about the two metrics since I believe I understand the point of what this is measuring a bit better.

alculquicondor replied:
@damemi, you missed the name change `sorting_deletion_age_ratio`

damemi replied:
thanks, updated

alculquicondor commented on pkg/controller/replicaset/replica_set.go:
```diff
@@ -802,10 +804,28 @@ func getPodsToDelete(filteredPods, relatedPods []*v1.Pod, diff int) []*v1.Pod {
 	if diff < len(filteredPods) {
 		podsWithRanks := getPodsRankedByRelatedPodsOnSameNode(filteredPods, relatedPods)
 		sort.Sort(podsWithRanks)
+		filteredPods = podsWithRanks.Pods
```
it doesn't seem like this line is necessary.

damemi replied:
I actually had a question about this line. Looking at the whole function, it seems like nothing is ever actually done with the sorted list of podRanks: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replicaset/replica_set.go#L799-L807

This is after [this change](https://github.com/kubernetes/kubernetes/commit/980b6406b24b41bc0ef061624385f1f420708819#diff-f658ce5cda23ed9920c4ee4eaaee968fb5a4dbd4a0ccc7f478db399f2b182380R762-R763), which added the call to `getPodsRankedByRelatedPodsOnSameNode`.

Prior to that, `sort()` was called directly on `filteredPods`: https://github.com/kubernetes/kubernetes/blame/865c3c56706d131c3164d16812fec58454a77020/pkg/controller/replicaset/replica_set.go#L684-L694

I asked @soltysh about this, and he wasn't sure either (though it seems odd that this sort of bug would go so long without being noticed).

Do you have any idea how the new `podsWithRanks` variable declared on L805 gets passed to the return statement at the bottom of the function?

damemi replied:
Ah maybe it is because `filteredPods` is the underlying slice in the returned `ActivePodsToRank` struct: https://github.com/kubernetes/kubernetes/blob/d8aad7944cc71314913fc0b43d6731e2c36f4928/pkg/controller/replicaset/replica_set.go#L813-L825

So calling `sort()` on the returned struct will modify `filteredPods` too?

alculquicondor replied:
correct :)

alculquicondor commented on pkg/controller/replicaset/replica_set.go:
```diff
@@ -802,10 +804,28 @@ func getPodsToDelete(filteredPods, relatedPods []*v1.Pod, diff int) []*v1.Pod {
 	return filteredPods[:diff]
 }
 
+func reportMetrics(filteredPods []*v1.Pod, diff int) {
```
`reportDeletedPodAgeRatioMetric`

alculquicondor commented on pkg/controller/replicaset/replica_set.go:
```diff
@@ -802,10 +804,28 @@ func getPodsToDelete(filteredPods, relatedPods []*v1.Pod, diff int) []*v1.Pod {
+	youngestTime := time.Time{}
+	// first we need to check all of the pods to get the youngest, as they may not necessarily be sorted by timestamp alone
+	for _, pod := range filteredPods {
+		if time.Since(pod.CreationTimestamp.Time) < time.Since(youngestTime) {
```
we are using a different "now" every time we call `time.Since`. Let's fix it.

alculquicondor commented on pkg/controller/replicaset/replica_set.go:
```diff
@@ -802,10 +804,28 @@ func getPodsToDelete(filteredPods, relatedPods []*v1.Pod, diff int) []*v1.Pod {
+
+	// for each pod chosen for deletion, report the ratio of its age to the youngest pod's age
+	for _, pod := range filteredPods[:diff] {
+		ratio := time.Since(pod.CreationTimestamp.Time).Seconds() / time.Since(youngestTime).Seconds()
```
this could very well be a zero denominator, as `pod.CreationTimestamp` has granularity of seconds :(

damemi replied:
Switching this to a single `time.Now` as mentioned above and using milliseconds should fix that right?

logicalhan commented:
/assign
/triage accepted

logicalhan commented on pkg/controller/replicaset/metrics/metrics.go:
```diff
@@ -0,0 +1,43 @@
+	&metrics.GaugeOpts{
+		Subsystem: ReplicaSetControllerSubsystem,
+		Name:      "deleted_pod_age_ratio",
+		Help:      "The ratio of chosen deleted pod's ages to the current youngest pod's age (at the time). Should be <2",
```
I find this metric incredibly confusing..

dgrisonnet replied:
I agree, without reading your [comment](https://github.com/kubernetes/kubernetes/pull/101767/files#r657268235) on the other thread, I had no clue what this metric was supposed to be used for just from reading the help text.

brancz commented:
I was talking to @logicalhan and we were wondering whether separate metrics had been considered in the KEP (I skimmed it but couldn't find it). We feel it might make it more understandable if we exposed the raw data as opposed to a calculated value. Typically raw values are preferred as we can change our mind about the calculations and it also happens more often than not that calculations contain some sort of bug and then historic data is incorrect.

brancz commented on pkg/controller/replicaset/metrics/metrics.go:
```diff
@@ -0,0 +1,42 @@
+
+const ReplicaSetControllerSubsystem = "replicaset_controller"
+
+var DeletedPodAgeRatio = metrics.NewGauge(
```
Agreed, after reading the KEP I expected histograms.

damemi reviewed (commented):
It sounds like we have reached an agreement on the current metric, so I pushed an update (as a new commit for now to preserve the review comments before squashing) that does:
* switch metric to histogram
* only report on ready pods

I think this was the only remaining feedback, is there anything I missed? If this looks good I will squash down

dgrisonnet commented on pkg/controller/replicaset/metrics/metrics.go:
```diff
@@ -0,0 +1,43 @@
+const ReplicaSetControllerSubsystem = "replicaset_controller"
+
+var DeletedPodAgeRatio = metrics.NewHistogram(
+	&metrics.HistogramOpts{
```
Do you really want to use the default buckets for your use case? They are meant for web services latency, so I don't think they should be a good fit for your use case.

damemi replied:
Good point, should I just manually define these buckets then as something like `[0.0, 2.0]`? If we are trying to measure the amount under `2`, does that make sense? I'm just not clear on the code syntax for defining these buckets, and couldn't find any good docs on it

dgrisonnet replied:
If the sole purpose of this metric is to measure all the operations done in less than 2s, then I don't think a histogram is the best fit. The best option for this case would be to use a counter accounting for all the operations that were completed in under 2 seconds. However, if you need more granularity on the duration of the operation, then a histogram will be a good fit.

As for defining custom buckets, you will need to think about the different time ranges that are meaningful to be collected for your use case.  For instance, the apiserver admission controller/webhook latency metrics are using these [custom buckets](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go#L48-L49).

alculquicondor replied:
The metric is about checking how far from 2s we get. Then, the range should be something like `[0, 8]`.

damemi replied:
So do we want several buckets on either side of 2? Like `[0, 0.5, 1, 1.5, 2, 2.5, 3, 4, 6, 8]`?

alculquicondor replied:
Can we use exponential buckets? I would start at 0.25 with a factor of 2.

dgrisonnet replied:
Yes, from an instrumentation perspective, I think it would make sense to have [exponential buckets](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics/histogram.go#L33-L36) here. Starting from 0.25 with a factor of 2, until you reach 8, so you will have a count of 6 and end up with the following buckets `[0.25, 0.5, 1, 2, 4, 8]`.

damemi replied:
Ok, thanks for the help everyone. I have pushed with the new buckets

alculquicondor commented on pkg/controller/replicaset/replica_set.go:
```diff
@@ -802,10 +804,31 @@ func getPodsToDelete(filteredPods, relatedPods []*v1.Pod, diff int) []*v1.Pod {
+func reportDeletedPodAgeRatioMetric(filteredPods []*v1.Pod, diff int) {
+	now := time.Now()
+	youngestTime := time.Time{}
+	// first we need to check all of the pods to get the youngest, as they may not necessarily be sorted by timestamp alone
```
all the ready pods.

alculquicondor reviewed (commented):
lgtm, other than the metric name.

dgrisonnet commented on pkg/controller/replicaset/metrics/metrics.go:
```diff
@@ -0,0 +1,43 @@
+var registerMetrics sync.Once
+
+// Register registers ReplicaSet controller metrics.
+func Register() {
```
Can we pass the registry as an argument here in order to prepare for an eventual decoupling of the global registry at some point? An example can be found here:

Register function: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/controllers/status/metrics.go#L64-L80

Registration: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go#L214-L221

damemi replied:
Does this need the `customRegistrationFunc` parameter too, with a `collector`? Or is it sufficient to just pass the registration func?

dgrisonnet replied:
The registration function is sufficient here, you don't need to pass a custom one.

damemi committed:
Promote LogarithmicScaleDown to beta

This promotes the LogarithmicScaleDown feature gate to Beta, enabling it
by default. It also introduces a new metric, `sorting_deletion_age_ratio`,
intended to measure the efficacy of this new replica set scaledown behavior.

dgrisonnet commented on pkg/controller/replicaset/metrics/metrics.go:
```diff
@@ -0,0 +1,47 @@
+		Help: "The ratio of chosen deleted pod's ages to the current youngest pod's age (at the time). Should be <2." +
+			"The intent of this metric is to measure the rough efficacy of the LogarithmicScaleDown feature gate's effect on" +
+			"the sorting (and deletion) of pods when a replicaset scales down",
+		Buckets: metrics.ExponentialBuckets(0.25, 2, 8),
```
```suggestion
		Buckets: metrics.ExponentialBuckets(0.25, 2, 6),
```
If you want the higher bound to be `8`, you need to set the number of buckets which is the last argument to 6. You will end up with 6 buckets `[0.25, 0.5, 1, 2, 4, 8]`.

damemi replied:
Ah yup, that's a typo on my part. Updated

dgrisonnet commented on pkg/controller/replicaset/replica_set.go:
```diff
@@ -112,6 +114,7 @@ func NewReplicaSetController(rsInformer appsinformers.ReplicaSetInformer, podInf
 	eventBroadcaster := record.NewBroadcaster()
 	eventBroadcaster.StartStructuredLogging(0)
 	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+	metrics.Register(legacyregistry.Register)
```
The registration must be done only once since we are still using the global registry and metrics can't be registered twice into the same registry. We should also check for errors returned by `Register()`.

We could also add a TODO to remember to replace the global registry at some point, as in: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go#L214-L221

damemi replied:
This should only be called once in this spot right, at controller initialization? Or is there somewhere better fit

logicalhan replied:
> The registration must be done only once since we are still using the global registry and metrics can't be registered twice into the same registry. We should also check for errors returned by `Register()`.

`Register` is resilient to registration multiple times, the error returned is generally "this metric is already registered", so it should be safe to call multiple times. The real danger is if there are more than one instances of the metric with the same identifier since the register would only have a reference to the first one registered. In that hypothetical case, the other identical but different instances of the metric would never be exposed via this metric endpoint.

dgrisonnet replied:
Yes it is resilient, but in practice, if we call `NewReplicaSetController` twice, the first call will register the metric families to the global registry, and the second will error since the metric families were already registered in the global registry.
Unless we move to a non-global registry, we will keep on logging registration errors whenever a new replicaset controller is created. To prevent that, I think we should add a global `sync.Once` that would make sure that the metrics are only registered once into the global registry.

logicalhan replied:
yeah i think that makes sense, but even with a non-global registry, we'd still be reregistering the metric (though i guess it could depend on how it was plumbed).

dgrisonnet replied:
yeah, I was thinking of having one registry per replicaset controller when commenting, but as you said, it depends on how it is implemented.

soltysh replied:
Honestly I can't think of a place where you'd call RS controller twice, and if you do you're in a bigger trouble b/c none of our controllers is resistant to that :sweat_smile: I'd suggest to address this as a followup PR to all controllers.

soltysh commented on pkg/controller/replicaset/metrics/metrics.go:
```diff
@@ -0,0 +1,43 @@
+var SortingDeletionAgeRatio = metrics.NewHistogram(
+	&metrics.HistogramOpts{
+		Subsystem: ReplicaSetControllerSubsystem,
+		Name:      "sorting_deletion_age_ratio",
```
@damemi would be good to update the KEP with this change.

damemi replied:
Opened here: https://github.com/kubernetes/enhancements/pull/2805

soltysh commented on pkg/controller/replicaset/metrics/metrics.go:
```diff
@@ -0,0 +1,43 @@
+		Help: "The ratio of chosen deleted pod's ages to the current youngest pod's age (at the time). Should be <2." +
+			"The intent of this metric is to measure the rough efficacy of the LogarithmicScaleDown feature gate's effect on" +
+			"the sorting (and deletion) of pods when a replicaset scales down",
+		Buckets: metrics.ExponentialBuckets(0.25, 2, 6),
```
I'd suggest marking the metric as `StabilityLevel` ALPHA for now with move to STABLE when GA-ing.

soltysh reviewed (approved):
A few nits & rebase and this is ready to go. 
/approve

soltysh commented:
/priority backlog

damemi commented:
Rebased and updated with @soltysh's comments

soltysh reviewed (approved):
this lgtm
leaving final tag to @logicalhan or @wojtek-t

wojtek-t commented on pkg/controller/replicaset/replica_set.go:
```diff
@@ -112,6 +114,10 @@ func NewReplicaSetController(rsInformer appsinformers.ReplicaSetInformer, podInf
 	eventBroadcaster.StartStructuredLogging(0)
 	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+	err := metrics.Register(legacyregistry.Register)
+	if err != nil {
```
nit:
```
if err := metrics.Register(...); err != nil {
 ...
```

wojtek-t commented on pkg/controller/replicaset/metrics/metrics.go:
```diff
@@ -0,0 +1,44 @@
+	&metrics.HistogramOpts{
+		Subsystem: ReplicaSetControllerSubsystem,
+		Name:      "sorting_deletion_age_ratio",
+		Help: "The ratio of chosen deleted pod's ages to the current youngest pod's age (at the time). Should be <2." +
```
Mention that we only take into account ready pods.

wojtek-t reviewed (commented):
OK - it's not perfect, but for the sake of making progress I'm fine what we have now modulo  smaller comments I added.

damemi commented:
rebased and updated with @wojtek-t's latest comments

soltysh reviewed (approved):
/lgtm

alculquicondor commented:
/retest

alculquicondor commented:
I suppose the failures are related to #103440, which should be fixed soon (PR is in the merge queue)
